modules/migrations/onedev.go

```diff
@@ -0,0 +1,562 @@
+	"code.gitea.io/gitea/modules/migrations/base"
+	"code.gitea.io/gitea/modules/structs"
+
+	jsoniter "github.com/json-iterator/go"
+)
+
+var (
+	_ base.Downloader        = &OneDevDownloader{}
+	_ base.DownloaderFactory = &OneDevDownloaderFactory{}
+)
+
+func init() {
+	RegisterDownloaderFactory(&OneDevDownloaderFactory{})
+}
+
+// OneDevDownloaderFactory defines a downloader factory
+type OneDevDownloaderFactory struct {
+}
+
+// New returns a downloader related to this factory according MigrateOptions
+func (f *OneDevDownloaderFactory) New(ctx context.Context, opts base.MigrateOptions) (base.Downloader, error) {
+	u, err := url.Parse(opts.CloneAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	repoName := ""
+
+	fields := strings.Split(strings.Trim(u.Path, "/"), "/")
+	if len(fields) == 2 && fields[0] == "projects" {
+		repoName = fields[1]
+	} else if len(fields) == 1 {
+		repoName = fields[0]
+	} else {
+		return nil, fmt.Errorf("invalid path: %s", u.Path)
+	}
+
+	u.Path = ""
+	u.Fragment = ""
+
+	log.Trace("Create onedev downloader. BaseURL: %v RepoName: %s", u, repoName)
+
+	return NewOneDevDownloader(ctx, u, opts.AuthUsername, opts.AuthPassword, repoName), nil
+}
+
+// GitServiceType returns the type of git service
+func (f *OneDevDownloaderFactory) GitServiceType() structs.GitServiceType {
+	return structs.OneDevService
+}
+
+type onedevUser struct {
+	ID    int64  `json:"id"`
+	Name  string `json:"name"`
+	Email string `json:"email"`
+}
+
+// OneDevDownloader implements a Downloader interface to get repository informations
+// from OneDev
+type OneDevDownloader struct {
+	base.NullDownloader
+	ctx           context.Context
+	client        *http.Client
+	baseURL       *url.URL
+	repoName      string
+	repoID        int64
+	maxIssueIndex int64
+	userMap       map[int64]*onedevUser
+	milestoneMap  map[int64]string
+}
+
+// SetContext set context
+func (d *OneDevDownloader) SetContext(ctx context.Context) {
+	d.ctx = ctx
+}
+
+// NewOneDevDownloader creates a new downloader
+func NewOneDevDownloader(ctx context.Context, baseURL *url.URL, username, password, repoName string) *OneDevDownloader {
+	var downloader = &OneDevDownloader{
+		ctx:      ctx,
+		baseURL:  baseURL,
+		repoName: repoName,
+		client: &http.Client{
+			Transport: &http.Transport{
+				Proxy: func(req *http.Request) (*url.URL, error) {
+					if len(username) > 0 && len(password) > 0 {
+						req.SetBasicAuth(username, password)
+					}
+					return nil, nil
+				},
+			},
+		},
+		userMap:      make(map[int64]*onedevUser),
+		milestoneMap: make(map[int64]string),
+	}
+
+	return downloader
+}
+
+func (d *OneDevDownloader) callAPI(endpoint string, parameter map[string]string, result interface{}) error {
+	u, err := d.baseURL.Parse(endpoint)
+	if err != nil {
+		return err
+	}
+
+	if parameter != nil {
+		query := u.Query()
+		for k, v := range parameter {
+			query.Set(k, v)
+		}
+		u.RawQuery = query.Encode()
+	}
+
+	req, err := http.NewRequest("GET", u.String(), nil)
```

Comment:
You must set the ctx on the request otherwise it is not cancellable.

```suggestion
	req, err := http.NewRequestWithContext(d.ctx, "GET", u.String(), nil)
```